File: javascript/04_streaming_response.js
/**
 * 04 - Streaming Response
 *
 * This example demonstrates real-time streaming of AI responses using
 * Server-Sent Events (SSE). Instead of waiting for the complete response,
 * tokens are received and displayed progressively as they're generated.
 *
 * Model used: gpt-5-nano (Fast OpenAI model)
 */

import OpenAI from 'openai';

// Initialize the Hibana client
const client = new OpenAI({
    apiKey: "YOUR_API_KEY",
    baseURL: "https://api-ai.hibanacloud.com/v1"
});

async function streamingChat() {
    /**
     * Demonstrate streaming responses with real-time output
     */

    console.log("=".repeat(60));
    console.log("Streaming Response Demo");
    console.log("=".repeat(60));

    const userMessage = "Write a short story about a robot learning to paint. Make it about 150 words.";

    console.log(`\nUser: ${userMessage}`);
    process.stdout.write("\nAssistant (streaming): ");

    // Create streaming chat completion
    const stream = await client.chat.completions.create({
        model: "gpt-5-nano",
        messages: [
            {
                role: "user",
                content: userMessage
            }
        ],
        temperature: 1.0,
        max_tokens: 10000,
        stream: true  // Enable streaming
    });

    // Collect the full response for later display
    let fullResponse = "";

    // Process the stream
    for await (const chunk of stream) {
        // Check if chunk has choices and content
        if (chunk.choices && chunk.choices.length > 0) {
            const content = chunk.choices[0].delta?.content;
            if (content) {
                fullResponse += content;
                // Print the content immediately (streaming effect)
                process.stdout.write(content);
            }
        }
    }

    console.log("\n\n" + "=".repeat(60));
    console.log("Streaming complete!");
    console.log("=".repeat(60));
    console.log(`\nTotal characters received: ${fullResponse.length}`);
}

async function streamingWithMetadata() {
    /**
     * Streaming with detailed chunk inspection
     */

    console.log("\n" + "=".repeat(60));
    console.log("Streaming with Metadata Inspection");
    console.log("=".repeat(60));

    const userMessage = "Explain quantum entanglement in 2 sentences.";

    console.log(`\nUser: ${userMessage}`);
    console.log("\nStreaming response...\n");

    const stream = await client.chat.completions.create({
        model: "gemini-2.5-flash-lite",  // Using Gemini model
        messages: [{ role: "user", content: userMessage }],
        temperature: 0.7,
        max_tokens: 10000,
        stream: true
    });

    let fullText = "";
    let chunkCount = 0;

    for await (const chunk of stream) {
        chunkCount++;

        // Inspect chunk structure
        if (chunk.choices && chunk.choices.length > 0) {
            const delta = chunk.choices[0].delta;

            // Check for content
            if (delta.content) {
                fullText += delta.content;
                process.stdout.write(delta.content);
            }

            // Check for finish reason
            if (chunk.choices[0].finish_reason) {
                console.log(`\n\nFinish reason: ${chunk.choices[0].finish_reason}`);
            }
        }
    }

    console.log(`\n\nTotal chunks received: ${chunkCount}`);
    console.log(`Full response length: ${fullText.length} characters`);
}

async function compareStreamingVsNormal() {
    /**
     * Compare streaming vs non-streaming response times
     */

    console.log("\n" + "=".repeat(60));
    console.log("Comparison: Streaming vs Non-Streaming");
    console.log("=".repeat(60));

    const message = "List 5 benefits of cloud computing.";

    // Non-streaming
    console.log("\n1. NON-STREAMING:");
    const startTime = Date.now();
    const response = await client.chat.completions.create({
        model: "gpt-5-nano",
        messages: [{ role: "user", content: message }],
        max_tokens: 10000,
        stream: false
    });
    const endTime = Date.now();

    console.log(`   Time to first output: ${((endTime - startTime) / 1000).toFixed(2)} seconds`);
    console.log(`   Response: ${response.choices[0].message.content.substring(0, 100)}...`);

    // Streaming
    console.log("\n2. STREAMING:");
    const startTimeStream = Date.now();
    let firstChunkTime = null;

    const stream = await client.chat.completions.create({
        model: "gpt-5-nano",
        messages: [{ role: "user", content: message }],
        max_tokens: 10000,
        stream: true
    });

    let i = 0;
    for await (const chunk of stream) {
        if (chunk.choices && chunk.choices.length > 0 && chunk.choices[0].delta?.content) {
            if (i === 0) {
                firstChunkTime = Date.now();
                console.log(`   Time to first output: ${((firstChunkTime - startTimeStream) / 1000).toFixed(2)} seconds`);
                console.log(`   First chunk: ${chunk.choices[0].delta.content}`);
                break;
            }
            i++;
        }
    }

    console.log("\n   Streaming provides faster time-to-first-token!");
}

// Run the examples
(async () => {
    try {
        // Example 1: Basic streaming
        await streamingChat();

        // Example 2: Streaming with metadata
        await streamingWithMetadata();

        // Example 3: Compare streaming vs non-streaming
        await compareStreamingVsNormal();

    } catch (error) {
        console.error(`\nError: ${error.message}`);
    }
})();

File: python/04_streaming_response.py
"""
04 - Streaming Response

This example demonstrates real-time streaming of AI responses using
Server-Sent Events (SSE). Instead of waiting for the complete response,
tokens are received and displayed progressively as they're generated.

Model used: gpt-5-nano (Fast OpenAI model)
"""

from openai import OpenAI
import sys

# Initialize the Hibana client
client = OpenAI(
    api_key="YOUR_API_KEY",
    base_url="https://api-ai.hibanacloud.com/v1"
)

def streaming_chat():
    """Demonstrate streaming responses with real-time output"""

    print("="*60)
    print("Streaming Response Demo")
    print("="*60)

    user_message = "Write a short story about a robot learning to paint. Make it about 150 words."

    print(f"\nUser: {user_message}")
    print("\nAssistant (streaming): ", end="", flush=True)

    # Create streaming chat completion
    stream = client.chat.completions.create(
        model="gpt-5-nano",
        messages=[
            {
                "role": "user",
                "content": user_message
            }
        ],
        temperature=1.0,
        max_tokens=10000,
        stream=True  # Enable streaming
    )

    # Collect the full response for later display
    full_response = ""

    # Process the stream
    for chunk in stream:
        # Check if chunk has choices and content
        if chunk.choices and len(chunk.choices) > 0:
            if chunk.choices[0].delta.content is not None:
                content = chunk.choices[0].delta.content
                full_response += content

                # Print the content immediately (streaming effect)
                print(content, end="", flush=True)

    print("\n\n" + "="*60)
    print("Streaming complete!")
    print("="*60)
    print(f"\nTotal characters received: {len(full_response)}")

def streaming_with_metadata():
    """Streaming with detailed chunk inspection"""

    print("\n" + "="*60)
    print("Streaming with Metadata Inspection")
    print("="*60)

    user_message = "Explain quantum entanglement in 2 sentences."

    print(f"\nUser: {user_message}")
    print("\nStreaming response...\n")

    stream = client.chat.completions.create(
        model="gemini-2.5-flash-lite",  # Using Gemini model
        messages=[{"role": "user", "content": user_message}],
        temperature=0.7,
        max_tokens=10000,
        stream=True
    )

    full_text = ""
    chunk_count = 0

    for chunk in stream:
        chunk_count += 1

        # Inspect chunk structure
        if chunk.choices and len(chunk.choices) > 0:
            delta = chunk.choices[0].delta

            # Check for content
            if delta.content:
                full_text += delta.content
                print(delta.content, end="", flush=True)

            # Check for finish reason
            if chunk.choices[0].finish_reason:
                print(f"\n\nFinish reason: {chunk.choices[0].finish_reason}")

    print(f"\n\nTotal chunks received: {chunk_count}")
    print(f"Full response length: {len(full_text)} characters")

def compare_streaming_vs_normal():
    """Compare streaming vs non-streaming response times"""

    import time

    print("\n" + "="*60)
    print("Comparison: Streaming vs Non-Streaming")
    print("="*60)

    message = "List 5 benefits of cloud computing."

    # Non-streaming
    print("\n1. NON-STREAMING:")
    start_time = time.time()
    response = client.chat.completions.create(
        model="gpt-5-nano",
        messages=[{"role": "user", "content": message}],
        max_tokens=10000,
        stream=False
    )
    end_time = time.time()

    print(f"   Time to first output: {end_time - start_time:.2f} seconds")
    print(f"   Response: {response.choices[0].message.content[:100]}...")

    # Streaming
    print("\n2. STREAMING:")
    start_time = time.time()
    first_chunk_time = None

    stream = client.chat.completions.create(
        model="gpt-5-nano",
        messages=[{"role": "user", "content": message}],
        max_tokens=10000,
        stream=True
    )

    for i, chunk in enumerate(stream):
        if chunk.choices and len(chunk.choices) > 0:
            if chunk.choices[0].delta.content:
                first_chunk_time = time.time()
                print(f"   Time to first output: {first_chunk_time - start_time:.2f} seconds")
                print(f"   First chunk: {chunk.choices[0].delta.content}")
                break

    print("\n   Streaming provides faster time-to-first-token!")

if __name__ == "__main__":
    try:
        # Example 1: Basic streaming
        streaming_chat()

        # Example 2: Streaming with metadata
        streaming_with_metadata()

        # Example 3: Compare streaming vs non-streaming
        compare_streaming_vs_normal()

    except Exception as e:
        print(f"\nError: {e}")
